src/Init/Data/BitVec/Lemmas.lean

```diff
@@ -789,6 +815,16 @@ protected theorem extractLsb_ofNat (x n : Nat) (hi lo : Nat) :
     getLsbD (extractLsb hi lo x) i = (i ≤ (hi-lo) && getLsbD x (lo+i)) := by
   simp [getLsbD, Nat.lt_succ]
 
+@[simp] theorem getMsbD_extract {hi lo : Nat} {x : BitVec w} {i : Nat} :
+    (extractLsb hi lo x).getMsbD i =
+    (decide (i < hi - lo + 1) && (decide (lo + (hi - lo - i) < w) && x.getMsbD (w - 1 - (lo + (hi - lo - i))))) := by
+  rw [BitVec.extractLsb, BitVec.getMsbD_extractLsb', Nat.add_sub_cancel]
+
+@[simp] theorem msb_extract {hi lo : Nat} {x : BitVec n} :
+    (extractLsb hi lo x).msb = (decide (0 < hi - lo + 1) &&
+    (decide (lo + (hi - lo) < n) && x.getMsbD (n - 1 - (lo + (hi - lo))))) := by
```

Comment:
```lean
    (extractLsb hi lo x).msb = (decide (0 < hi - lo + 1) &&
    (decide (lo + (hi - lo) < n) && x.getMsbD (n - 1 - (lo + (hi - lo))))) := by
```

I feel the conditions can be simplified. We should be able to write `decide (0 < hi - lo + 1)` as `decide (lo <= hi)`. Similarly, given this, is it not the case that `decide (lo + (hi - lo) < n)` is equivalent to `decide (hi < n)`, and the final case is `x.getMsbD (n - 1 -  hi)`? Note that I'm not 100% sure about the above computations, but it does seem to me that `lo + (hi - lo)` is refactorable by using another `decide (...)` condition if necessary

Reply:
@luisacicolini, it's helpful if you can click "resolve conversation" for resolved conversations, so reviewers don't have to evaluate the status of the conversation themselves.